tsdb/querier.go

```diff
@@ -375,38 +375,31 @@ func inversePostingsForMatcher(ix IndexReader, m *labels.Matcher) (index.Posting
 }
 
 func labelValuesWithMatchers(r IndexReader, name string, matchers ...*labels.Matcher) ([]string, error) {
-	// We're only interested in metrics which have the label <name>.
-	requireLabel, err := labels.NewMatcher(labels.MatchNotEqual, name, "")
+	p, err := PostingsForMatchers(r, matchers...)
 	if err != nil {
-		return nil, errors.Wrapf(err, "Failed to instantiate label matcher")
+		return nil, errors.Wrap(err, "fetching postings for matchers")
 	}
 
-	var p index.Postings
-	p, err = PostingsForMatchers(r, append(matchers, requireLabel)...)
+	allValues, err := r.LabelValues(name)
 	if err != nil {
-		return nil, err
+		return nil, errors.Wrapf(err, "fetching values of label %s", name)
 	}
-
-	dedupe := map[string]interface{}{}
-	for p.Next() {
-		v, err := r.LabelValueFor(p.At(), name)
+	valuesPostings := make([]index.Postings, len(allValues))
+	for i, value := range allValues {
+		valuesPostings[i], err = r.Postings(name, value)
 		if err != nil {
-			if err == storage.ErrNotFound {
-				continue
-			}
-
-			return nil, err
+			return nil, errors.Wrapf(err, "fetching postings for %s=%q", name, value)
 		}
-		dedupe[v] = nil
 	}
-
-	if err = p.Err(); err != nil {
-		return nil, err
+	indexes, err := index.FindIntersectingPostings(p, valuesPostings)
+	if err != nil {
+		return nil, errors.Wrap(err, "intersecting postings")
 	}
+	sort.Ints(indexes)
```

Comment:
We don't need it sorted (we were not sorting `values` before)

Reply:
Oh right, removed in https://github.com/prometheus/prometheus/pull/9907/commits/b0143cb602b315dafbfa464199fc007971727a00